FAILED NewConnectionPoolSpec.The host-level client infrastructure should never close hot connections when minConnections key is given and >0 (minConnections = 5)

https://jenkins.akka.io:8498/job/akka-http-nightly-against-akka-2.5-on-2.12/8/testReport/akka.http.impl.engine.client/NewConnectionPoolSpec/The_host_level_client_infrastructure_should_never_close_hot_connections_when_minConnections_key_is_given_and__0__minConnections___5_/

```
Error Message

assertion failed: timeout (5 seconds) during expectMsgClass waiting for class akka.stream.testkit.TestSubscriber$OnNext

Stacktrace

java.lang.AssertionError: assertion failed: timeout (5 seconds) during expectMsgClass waiting for class akka.stream.testkit.TestSubscriber$OnNext
	at scala.Predef$.assert(Predef.scala:219)
	at akka.testkit.TestKitBase.expectMsgClass_internal(TestKit.scala:508)
	at akka.testkit.TestKitBase.expectMsgType(TestKit.scala:481)
	at akka.testkit.TestKitBase.expectMsgType$(TestKit.scala:481)
	at akka.testkit.TestKit.expectMsgType(TestKit.scala:882)
	at akka.stream.testkit.TestSubscriber$ManualProbe.expectNextN(StreamTestKit.scala:374)
	at akka.http.impl.engine.client.ConnectionPoolSpec$$anon$17.$anonfun$new$23(ConnectionPoolSpec.scala:341)
	at akka.http.impl.engine.client.ConnectionPoolSpec$$anon$17.$anonfun$new$23$adapted(ConnectionPoolSpec.scala:336)
	at scala.collection.immutable.Range.foreach(Range.scala:156)
	at akka.http.impl.engine.client.ConnectionPoolSpec$$anon$17.<init>(ConnectionPoolSpec.scala:336)
	at akka.http.impl.engine.client.ConnectionPoolSpec.$anonfun$new$22(ConnectionPoolSpec.scala:326)
```

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala
@@ -148,6 +148,7 @@ private[client] object NewHostConnectionPool {
           private[this] var state: SlotState = SlotState.Unconnected
           private[this] var currentTimeoutId: Long = -1
           private[this] var currentTimeout: Cancellable = _
+          private[this] var isEnqueuedForResponseDispatch: Boolean = false
 
           private[this] var connection: SlotConnection = _
           def isIdle: Boolean = state.isIdle
@@ -179,8 +180,11 @@ private[client] object NewHostConnectionPool {
 
           def onResponseReceived(response: HttpResponse): Unit =
             updateState(Event.onResponseReceived, response)
-          def onResponseDispatchable(): Unit =
+
+          def onResponseDispatchable(): Unit = {
+            isEnqueuedForResponseDispatch = false
             updateState(Event.onResponseDispatchable)
+          }
 
           def onResponseEntitySubscribed(): Unit =
             updateState(Event.onResponseEntitySubscribed)
@@ -232,7 +236,10 @@ private[client] object NewHostConnectionPool {
                   case _: WaitingForResponseDispatch ⇒
                     if (isAvailable(responsesOut)) OptionVal.Some(Event.onResponseDispatchable)
                     else {
-                      logic.slotsWaitingForDispatch.addLast(this)
+                      if (!isEnqueuedForResponseDispatch) {
+                        isEnqueuedForResponseDispatch = true
+                        logic.slotsWaitingForDispatch.addLast(this)
+                      }
                       OptionVal.None
                     }
                   case WaitingForResponseEntitySubscription(_, HttpResponse(_, _, _: HttpEntity.Strict, _), _) ⇒

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala
@@ -17,12 +17,11 @@ import akka.http.impl.util.{ RichHttpRequest, StageLoggingWithOverride, StreamUt
 import akka.http.scaladsl.Http
 import akka.http.scaladsl.model.{ HttpEntity, HttpRequest, HttpResponse, headers }
 import akka.http.scaladsl.settings.ConnectionPoolSettings
+import akka.stream._
 import akka.stream.scaladsl.{ Flow, Keep, Sink, Source }
 import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler, OutHandler }
-import akka.stream._
 import akka.util.OptionVal
 
-import scala.annotation.tailrec
 import scala.concurrent.Future
 import scala.concurrent.duration.FiniteDuration
 import scala.util.control.NonFatal
@@ -110,7 +109,7 @@ private[client] object NewHostConnectionPool {
             slots.find(_.isIdle)
               .getOrElse(throw new IllegalStateException("Tried to dispatch request when no slot is idle"))
 
-          slot.debug("Dispatching request") // FIXME: add abbreviation
+          slot.debug("Dispatching request [{}]", req.request.debugString)
           slot.onNewRequest(req)
         }
 

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala
@@ -111,12 +111,41 @@ private[client] object NewHostConnectionPool {
               .getOrElse(throw new IllegalStateException("Tried to dispatch request when no slot is idle"))
 
           slot.debug("Dispatching request") // FIXME: add abbreviation
-          slot.dispatchRequest(req)
+          slot.onNewRequest(req)
         }
 
         def numConnectedSlots: Int = slots.count(_.isConnected)
 
-        private class Slot(val slotId: Int) extends SlotContext {
+        final case class Event[T](name: String, transition: (SlotState, Slot, T) ⇒ SlotState) {
+          override def toString: String = s"Event($name)"
+        }
+        object Event {
+          val onPreConnect = event0("onPreConnect", _.onPreConnect(_))
+          val onConnectionAttemptSucceeded = event[Http.OutgoingConnection]("onConnectionAttemptSucceeded", _.onConnectionAttemptSucceeded(_, _))
+          val onConnectionAttemptFailed = event[Throwable]("onConnectionAttemptFailed", _.onConnectionAttemptFailed(_, _))
+          val onNewRequest = event[RequestContext]("onNewRequest", _.onNewRequest(_, _))
+
+          val onRequestEntityCompleted = event0("onRequestEntityCompleted", _.onRequestEntityCompleted(_))
+          val onRequestEntityFailed = event[Throwable]("onRequestEntityFailed", _.onRequestEntityFailed(_, _))
+
+          val onResponseReceived = event[HttpResponse]("onResponseReceived", _.onResponseReceived(_, _))
+          val onResponseDispatchable = event0("onResponseDispatchable", _.onResponseDispatchable(_))
+          val onResponseEntitySubscribed = event0("onResponseEntitySubscribed", _.onResponseEntitySubscribed(_))
+          val onResponseEntityCompleted = event0("onResponseEntityCompleted", _.onResponseEntityCompleted(_))
+          val onResponseEntityFailed = event[Throwable]("onResponseEntityFailed", _.onResponseEntityFailed(_, _))
+
+          val onConnectionCompleted = event0("onConnectionCompleted", _.onConnectionCompleted(_))
+          val onConnectionFailed = event[Throwable]("onConnectionFailed", _.onConnectionFailed(_, _))
+
+          val onTimeout = event0("onTimeout", _.onTimeout(_))
+
+          val setState = event[SlotState]("setState", (old, slot, newState) ⇒ newState)
+
+          private def event0(name: String, transition: (SlotState, Slot) ⇒ SlotState): Event[Unit] = new Event(name, (state, slot, _) ⇒ transition(state, slot))
+          private def event[T](name: String, transition: (SlotState, Slot, T) ⇒ SlotState): Event[T] = new Event[T](name, transition)
+        }
+
+        final class Slot(val slotId: Int) extends SlotContext {
           private[this] var state: SlotState = SlotState.Unconnected
           private[this] var currentTimeoutId: Long = -1
           private[this] var currentTimeout: Cancellable = _
@@ -132,49 +161,49 @@ private[client] object NewHostConnectionPool {
           }
 
           def initialize(): Unit =
-            if (slotId < settings.minConnections) {
-              debug("Preconnecting")
-              updateState(_.onPreConnect(this))
-            }
+            if (slotId < settings.minConnections)
+              updateState(Event.onPreConnect)
 
-          def onConnected(outgoing: Http.OutgoingConnection): Unit =
-            updateState(_.onConnectionAttemptSucceeded(this, outgoing))
+          def onConnectionAttemptSucceeded(outgoing: Http.OutgoingConnection): Unit =
+            updateState(Event.onConnectionAttemptSucceeded, outgoing)
 
-          def onConnectFailed(cause: Throwable): Unit =
-            updateState(_.onConnectionAttemptFailed(this, cause))
+          def onConnectionAttemptFailed(cause: Throwable): Unit =
+            updateState(Event.onConnectionAttemptFailed, cause)
 
-          def dispatchRequest(req: RequestContext): Unit =
-            updateState(_.onNewRequest(this, req))
+          def onNewRequest(req: RequestContext): Unit =
+            updateState(Event.onNewRequest, req)
 
           def onRequestEntityCompleted(): Unit =
-            updateState(_.onRequestEntityCompleted(this))
+            updateState(Event.onRequestEntityCompleted)
           def onRequestEntityFailed(cause: Throwable): Unit =
-            updateState(_.onRequestEntityFailed(this, cause))
+            updateState(Event.onRequestEntityFailed, cause)
 
           def onResponseReceived(response: HttpResponse): Unit =
-            updateState(_.onResponseReceived(this, response))
+            updateState(Event.onResponseReceived, response)
           def onResponseDispatchable(): Unit =
-            updateState(_.onResponseDispatchable(this))
+            updateState(Event.onResponseDispatchable)
 
           def onResponseEntitySubscribed(): Unit =
-            updateState(_.onResponseEntitySubscribed(this))
+            updateState(Event.onResponseEntitySubscribed)
           def onResponseEntityCompleted(): Unit =
-            updateState(_.onResponseEntityCompleted(this))
+            updateState(Event.onResponseEntityCompleted)
           def onResponseEntityFailed(cause: Throwable): Unit =
-            updateState(_.onResponseEntityFailed(this, cause))
+            updateState(Event.onResponseEntityFailed, cause)
 
-          def onConnectionCompleted(): Unit = updateState(_.onConnectionCompleted(this))
-          def onConnectionFailed(cause: Throwable): Unit = updateState(_.onConnectionFailed(this, cause))
+          def onConnectionCompleted(): Unit =
+            updateState(Event.onConnectionCompleted)
+          def onConnectionFailed(cause: Throwable): Unit =
+            updateState(Event.onConnectionFailed, cause)
 
-          type StateTransition = SlotState ⇒ SlotState
-          protected def updateState(f: StateTransition): Unit = {
-            def runOneTransition(f: StateTransition): OptionVal[StateTransition] =
+          protected def updateState(event: Event[Unit]): Unit = updateState(event, ())
+          protected def updateState[T](event: Event[T], arg: T): Unit = {
+            def runOneTransition[U](event: Event[U], arg: U): OptionVal[Event[Unit]] =
               try {
                 cancelCurrentTimeout()
 
                 val previousState = state
-                state = f(state)
-                debug(s"State change [${previousState.name}] -> [${state.name}]")
+                state = event.transition(state, this, arg)
+                debug(s"After event [${event.name}] State change [${previousState.name}] -> [${state.name}]")
 
                 state.stateTimeout match {
                   case d: FiniteDuration ⇒
@@ -184,7 +213,7 @@ private[client] object NewHostConnectionPool {
                       materializer.scheduleOnce(d, safeRunnable {
                         if (myTimeoutId == currentTimeoutId) { // timeout may race with state changes, ignore if timeout isn't current any more
                           debug(s"Slot timeout after $d")
-                          updateState(_.onTimeout(this))
+                          updateState(Event.onTimeout)
                         }
                       })
                   case _ ⇒ // no timeout set, nothing to do
@@ -202,20 +231,20 @@ private[client] object NewHostConnectionPool {
 
                 state match {
                   case _: WaitingForResponseDispatch ⇒
-                    if (isAvailable(responsesOut)) OptionVal.Some(_.onResponseDispatchable(this))
+                    if (isAvailable(responsesOut)) OptionVal.Some(Event.onResponseDispatchable)
                     else {
                       logic.slotsWaitingForDispatch.addLast(this)
                       OptionVal.None
                     }
                   case WaitingForResponseEntitySubscription(_, HttpResponse(_, _, _: HttpEntity.Strict, _), _) ⇒
                     // the connection cannot drive these for a strict entity so we have to loop ourselves
-                    OptionVal.Some(_.onResponseEntitySubscribed(this))
+                    OptionVal.Some(Event.onResponseEntitySubscribed)
                   case WaitingForEndOfResponseEntity(_, HttpResponse(_, _, _: HttpEntity.Strict, _)) ⇒
                     // the connection cannot drive these for a strict entity so we have to loop ourselves
-                    OptionVal.Some(_.onResponseEntityCompleted(this))
+                    OptionVal.Some(Event.onResponseEntityCompleted)
                   case Unconnected if numConnectedSlots < settings.minConnections ⇒
                     debug(s"Preconnecting because number of connected slots fell down to $numConnectedSlots")
-                    OptionVal.Some(_.onPreConnect(this))
+                    OptionVal.Some(Event.onPreConnect)
                   case _ ⇒ OptionVal.None
                 }
 
@@ -237,26 +266,27 @@ private[client] object NewHostConnectionPool {
                       error(ex, "Shutting down slot after error failed.")
                   }
                   state = Unconnected
-                  OptionVal.Some(_.onPreConnect(this))
+                  OptionVal.Some(Event.onPreConnect)
               }
 
             /** Run a loop of state transitions */
-            @tailrec def loop(f: StateTransition, remainingIterations: Int): Unit =
+            /* @tailrec (does not work for some reason?) */
+            def loop[U](event: Event[U], arg: U, remainingIterations: Int): Unit =
               if (remainingIterations > 0)
-                runOneTransition(f) match {
+                runOneTransition(event, arg) match {
                   case OptionVal.None       ⇒ // no more changes
-                  case OptionVal.Some(next) ⇒ loop(next, remainingIterations - 1)
+                  case OptionVal.Some(next) ⇒ loop(next, (), remainingIterations - 1)
                 }
               else
                 throw new IllegalStateException(
                   "State transition loop exceeded maximum number of loops. The pool will shutdown itself. " +
                     "That's probably a bug. Please file a bug at https://github.com/akka/akka-http/issues. ")
 
-            loop(f, 10)
+            loop(event, arg, 10)
           }
 
           protected def setState(newState: SlotState): Unit =
-            updateState(_ ⇒ newState)
+            updateState(Event.setState, newState)
 
           def debug(msg: String): Unit =
             if (log.isDebugEnabled)
@@ -439,8 +469,8 @@ private[client] object NewHostConnectionPool {
           responseIn.setHandler(slotCon)
 
           connection.onComplete(safely {
-            case Success(outgoingConnection) ⇒ slotCon.withSlot(_.onConnected(outgoingConnection))
-            case Failure(cause)              ⇒ slotCon.withSlot(_.onConnectFailed(cause))
+            case Success(outgoingConnection) ⇒ slotCon.withSlot(_.onConnectionAttemptSucceeded(outgoingConnection))
+            case Failure(cause)              ⇒ slotCon.withSlot(_.onConnectionAttemptFailed(cause))
           })(ExecutionContexts.sameThreadExecutionContext)
 
           slotCon

--- a/akka-http-core/src/test/scala/akka/http/impl/engine/client/HostConnectionPoolSpec.scala
+++ b/akka-http-core/src/test/scala/akka/http/impl/engine/client/HostConnectionPoolSpec.scala
@@ -337,6 +337,38 @@ class HostConnectionPoolSpec extends AkkaSpec(
           a[Throwable] should be thrownBy oneCycle()
         }
       }
+      "dispatch multiple failures on different slots when request entity fails" in new SetupWithServerProbes(_.withMaxConnections(3)) {
+        val req1 = pushChunkedRequest(numRetries = 0)
+        val conn1 = expectNextConnection()
+        val req1Bytes = conn1.expectChunkedRequestBytesAsProbe()
+
+        val req2 = pushChunkedRequest(numRetries = 0)
+        val conn2 = expectNextConnection()
+        val req2Bytes = conn2.expectChunkedRequestBytesAsProbe()
+
+        req1.sendError(new RuntimeException("First request stumbled and fell"))
+        conn1.failConnection(new RuntimeException("First connection crash-landed on mars"))
+
+        // don't check for first error yet
+
+        req2.sendError(new RuntimeException("Second request stumbled and flew"))
+        conn2.failConnection(new RuntimeException("Second connection crash-landed on the moon"))
+
+        expectResponseError()
+        expectResponseError()
+
+        // check that we are not still dispatchable (#1726)
+        pushRequest()
+        pushRequest()
+
+        val conn3 = expectNextConnection()
+        val conn4 = expectNextConnection()
+
+        conn3.pushResponse()
+        conn4.pushResponse()
+        expectResponse()
+        expectResponse()
+      }
       "not send requests to known-to-be-closed-soon connections" in pending
       "support retries" in pending
       "strictly enforce number of established connections in longer running case" in pending

